internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -90,73 +90,125 @@ func scenariosNamesToList(integrationTestScenarios *[]v1beta2.IntegrationTestSce
 	return &result
 }
 
-// EnsureRerunPipelineRunsExist is responsible for recreating integration test pipelines triggered by users
+// EnsureRerunPipelineRunsExist is responsible for recreating integration test pipelineruns triggered by users
 func (a *Adapter) EnsureRerunPipelineRunsExist() (controller.OperationResult, error) {
-
-	scenarioName, ok := gitops.GetIntegrationTestRunLabelValue(a.snapshot)
+	runLabelValue, ok := gitops.GetIntegrationTestRunLabelValue(a.snapshot)
 	if !ok {
-		// no test rerun triggered
 		return controller.ContinueProcessing()
 	}
-	integrationTestScenario, err := a.loader.GetScenario(a.context, a.client, scenarioName, a.application.Namespace)
 
+	integrationTestScenarios, opResult, err := a.getScenariosToRerun(runLabelValue)
+	if integrationTestScenarios == nil {
+		return opResult, err
+	}
+
+	testStatuses, err := gitops.NewSnapshotIntegrationTestStatusesFromSnapshot(a.snapshot)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	skipScenarioRerunCount, opResult, err := a.handleScenarioReruns(integrationTestScenarios, testStatuses)
+	if opResult.CancelRequest || err != nil {
+		return opResult, err
+	}
+
+	if err = a.cleanupRerunLabelAndUpdateStatus(skipScenarioRerunCount, len(*integrationTestScenarios), runLabelValue, testStatuses); err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	return controller.ContinueProcessing()
+}
+
+// getScenariosToRerun fetches and filters the IntegrationTestScenarios based on runLabelValue.
+func (a *Adapter) getScenariosToRerun(runLabelValue string) (*[]v1beta2.IntegrationTestScenario, controller.OperationResult, error) {
+	if runLabelValue == "all" {
+		scenarios, err := a.loader.GetAllIntegrationTestScenariosForApplication(a.context, a.client, a.application)
+		if err != nil {
+			a.logger.Error(err, "Failed to get IntegrationTestScenarios", "Application.Namespace", a.application.Namespace)
+			opResult, err := controller.RequeueWithError(err)
+			return nil, opResult, err
+		}
+		filteredScenarios := gitops.FilterIntegrationTestScenariosWithContext(scenarios, a.snapshot)
+		if filteredScenarios == nil {
+			a.logger.Info("None of the Scenarios' context are applicable to the Snapshot, nothing to re-run")
+		}
```

Comment:
Would the new [GetRequiredIntegrationTestScenariosForSnapshot](https://github.com/konflux-ci/integration-service/blob/7a6ff94f0ce3a7dc074f7382e4d5d6b78b278b3d/loader/loader.go#L227C18-L227C64) function be a good replacement for this code block?

Reply:
Hmm, that function only returns required ITS, but when a user puts the `run` label's value to "all" maybe they'd expect to run all the ITS (even optional ones), or would they not?

The answer to the above question would decide if we want to refactor the suggested function to have a boolean variable to deal with that.

Reply:
Oh, good point. Yeah maybe we could make that function a bit more generic by adding an `includeOptional` bool that dictates if it also returns the optional IntegrationTestScenarios. 
Also @hongweiliu17 wdyt?

Reply:
Tbh this is more of a nice-to have suggestion, so I'm also fine if the code remains as-is.

Reply:
Sorry, I didn't notice this message previously.

Reply:
I agree to have this nice-to-have, the function can the name like GetAllIntegrationTestScenariosForSnapshot(), then we can reduce the code complexity all the time.